### ISSUE TITLE
spring-cloud-sleuth and spring-cloud-stream-binder-kafka

### ISSUE THREAD

**Issue by marcin-kasinski** [spring-cloud/spring-cloud-sleuth#944]
Environment:

spring-boot-version: 2.0.1.RELEASE
spring-cloud.version:Finchley.M9
pom:https://github.com/marcin-kasinski/SpringBootWeb/blob/master/pom.xml
bootstrap.yml :https://github.com/marcin-kasinski/SpringBootWeb/blob/master/src/main/resources/bootstrap-prd.yml

I have simple web application.

It send messages to kafka service with spring-cloud-stream-binder-kafka.

It sends spans to zipkin with spring-cloud-sleuth-zipkin.

Problem is that my app tries to send spans via kafka not http with topic name zipkin.
I think it is because spring-cloud-stream-binder-kafka dependencies in my pom .

My questions:

1:

What should I do to send my spand via HTTP not KAFKA in my app

2.

Why when this app tries to send spans via kafka doesn't use my kafka konfiguration from bootstrap.yml but sends spand via kafka on localhost ?

What configuration parameters I shoud use to send spand via my kafka server not default localhost ?

Regards Marcin

**Comment by marcingrzejszczak** [comment]
Siemano ;)

> What should I do to send my spand via HTTP not KAFKA in my app

Have you tried searching the docs? https://cloud.spring.io/spring-cloud-sleuth/single/spring-cloud-sleuth.html#_sending_spans_to_zipkin

> Why when this app tries to send spans via kafka doesn't use my kafka konfiguration from bootstrap.yml but sends spand via kafka on localhost ?

It reuses the `KafkaProperties` bean. Just set the `bootstrap.servers` to change the server location (https://docs.spring.io/spring-kafka/reference/htmlsingle/).,

**Comment by marcin-kasinski** [comment]
Cześć. :)

I've added 

    zipkin:
        baseUrl: http://springbootzipkin:7777
        sender:
            type: WEB

to my bootstrap.yml web app and kafka listener  app


In web app logs I can see:

2018-04-16 10:29:42.770  INFO [SpringBootWeb,076beea38a2afb34,076beea38a2afb34,true] 16 --- [nio-7070-exec-7] mk.WebController                         : Message sent


In kafka listener  app I can see the same traceid

2018-04-16 10:29:42.984  INFO [SpringBootKafkaListener,076beea38a2afb34,dd990dc084d8a299,true] 16 --- [container-0-C-1] mk.KafkaApplicationListener              : orderIn.getId() 16 Apr 2018 10:29:42 GMT

Problem is that I can not see this records in zipkin dashboard.

I can see other records with the same trace ID (Web app send additional requests via REST). I can see only records from spring boot apps 1.5.10

This problem exists only with spring boot 2.0.1 and spring cloud Finchley.M9
With spring boot 1.5.10 and spring cloud Dalston.SR3 everything works fine.

Regards

**Comment by marcingrzejszczak** [comment]
The property is `spring.zipkin.sender.type: web` not `zipkin...` you haven't done a typo right?

Also, I'm confused. You're using Kafka to send a message to some other app, or you're using Kafka as a broker between your app and Zipkin?

**Comment by marcin-kasinski** [comment]
I've prepared simple spring cloud POT

I have Spring Cloud Sleuth simple scenario.

Spring boot web app (1.5.10) sends businness messages to spring boot kafka listener (1.5.10)   , spring boot rabbit listener (1.5.10) and spring boot REST service (1.5.10)

With these versions everything works OK. I have all records in zipkin dashboard

Then I've tried moved to 2.0.1

Spring boot web app (2.0.1) sends businness messages to spring boot kafka listener (2.0.1)   , spring boot rabbit listener (1.5.10) and spring boot REST service (1.5.10)


With this configuration my app works but in zipkin dashboard I can see only records from rabibit listener and rest service only (spring boot 1.5.10 ). There is no records from web app and kafka app which I moved to 2.0.1 and Finchley.M9.

After moved to 2.0.1 everything works fine except Spring cloud sleuth.
Is there any debug property which let me see zipkin diagnostic traces.

In logs I can see that log message is sampled by I can not see it in my zipkin server.

I think my configuration is OK.
You can check spring boot web conf here:

pom:https://github.com/marcin-kasinski/SpringBootWeb/blob/master/pom.xml
bootstrap.yml :https://github.com/marcin-kasinski/SpringBootWeb/blob/master/src/main/resources/bootstrap-prd.yml

**Comment by liugogal** [comment]
I have  same problem.  using 2.0.1 and Finchley.M9 .

2018-04-17 09:26:11.737 DEBUG [order,46c5f2118bff143a,98af63bd859055e4,true] 1199 --- [strix-product-2] o.s.c.openfeign.support.SpringEncoder    : Writing [[DecreaseStockInput(productId=164103465734242707, productQuantity=2)]] using [org.springframework.http.converter.json.MappingJackson2HttpMessageConverter@749c0669]

zipkin ui show noting

**Comment by marcingrzejszczak** [comment]
@marcin-kasinski you have only one application? The `SpringBootWeb` communicates with some other app?
@liugogal I have no idea what this log is supposed to tell me.

**Comment by marcingrzejszczak** [comment]
@marcin-kasinski your repo weights more than 80 MBs. Can you do sth about it please? Can you create a new sample or rewrite the history so that your fat jar is not here https://github.com/marcin-kasinski/SpringBootWeb/tree/master/docker

**Comment by marcingrzejszczak** [comment]
@marcin-kasinski I can't really replicate the problem. Can you write a script or a set of steps I need to do to replicate the issue? I see that Zipkin has to be in service discovery (I assume that you're running it in a Kubernetes cluster), I need Kafka etc. Please describe the steps.

**Comment by marcingrzejszczak** [comment]
also you've completely misunderstood what I wrote

```yml
    zipkin:
        baseUrl: http://springbootzipkin:7777
        sender:
            type: web
    sleuth:
        sampler:
            percentage: 1
            probability: 1
#if you want to send spans via kafka...            
    kafka:
        bootstrap-servers: kafka-service:9092 
```

here you state that you will send spans via WEB 

```yml
    zipkin:
        baseUrl: http://springbootzipkin:7777
        sender:
            type: web
```

yet here you state

```yml
        bootstrap-servers: kafka-service:9092 
```

which is completely untrue. If you want to send spans via Kafka you have to set `type: kafka` . 

I think I completely confused in what you're trying to achieve.

**Comment by marcin-kasinski** [comment]
In my conf there are lots of additional parameters .
I tought it doesn't matter.

You can check my new projest (https://github.com/marcin-kasinski/SpringBootWebTST).
In my opinion it si as simple as possible.

My aim is to send spans via web to zipkin.

How I've tested my web app:

1. Start zipkin on localhost:7777 (my version is 1.5.10.RELEASE and Dalston.SR3)
2. start my web project
3. In web browser open my web app http://localhost:7070/
4. Check in web app logs traceid
5. Find it in zipkin.

My example result: ERROR: Cannot load trace ea0d1c6a647b9d10: server error (error)

Then 

1. stop my web project

2. IN pom change 

  		<spring-cloud.version>Finchley.M9</spring-cloud.version>
 
  		<spring-boot-version>2.0.1.RELEASE</spring-boot-version>

to:

  		<spring-cloud.version>Dalston.SR3</spring-cloud.version>
 
  		<spring-boot-version>1.5.10.RELEASE</spring-boot-version>



3. start my web project
4. In web browser open my web app http://localhost:7070/
5. Check in web app logs traceid
6. Find it in zipkin.
7. Here with 1.5.10 I can find it in zipkin.

To summarise:

My simple web app with 1.5.10 and Dalston.SR3 works OK.
If I change only spring boot wersion to 2.0.1 and cloud version to Finchley.M9 it doesn't work.

Regards Marcin

**Comment by marcingrzejszczak** [comment]
You're using wrong dependencies

You have:

```xml
<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter-sleuth</artifactId>
		</dependency>
<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-sleuth-zipkin</artifactId>
		</dependency>
```

instead of 

```xml

<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter-zipkin</artifactId>
		</dependency>
```

I've fixed this entry, set up Zipkin at port 9411, updated Zipkin's base url to point to 9411 and I got the following:

![image](https://user-images.githubusercontent.com/3297437/38874167-c34faa1e-4257-11e8-8950-70247d0f6fdd.png)

**Comment by marcin-kasinski** [comment]
Kill me...

I've fixed my pom file but it still doesn't work.

Can you check my pom file again ?

**Comment by marcingrzejszczak** [comment]
I accidentally applied the change from spaces to tabs :|

**Comment by marcin-kasinski** [comment]
Can you send me pom.xml full file not patch please.
I have Hunk #3 FAILED and I want to be 100% sure we have the same files.
Wiem jestem lama :)

Regards Marcin

**Comment by marcingrzejszczak** [comment]
Zlituj sie... :P https://github.com/marcin-kasinski/SpringBootWebTST/pull/1/files

**Comment by marcin-kasinski** [comment]
Sorry.

I didn't noticed it was github diff.

I've tried to patch my local file.

Now it is working but it wasn;t my web app .

It was zipkin server ?

My zipkin server replied with 404 every time.

I've used my spring boot application with @EnableZipkinServer annotation .

With https://zipkin.io/quickstart.sh it works ok.
